File: Java/playwright-tests/src/main/java/com/qa/portfolio/pages/InventoryPage.java
package com.qa.portfolio.pages;

import com.microsoft.playwright.Page;

import java.util.List;

public class InventoryPage extends BasePage {

    private static final String INVENTORY_LIST  = ".inventory_list";
    private static final String ITEM_NAMES      = ".inventory_item_name";
    private static final String ADD_TO_CART_BTN = ".inventory_item button[id^='add-to-cart']";
    private static final String REMOVE_BTN      = ".inventory_item button[id^='remove']";
    private static final String CART_BADGE      = ".shopping_cart_badge";
    private static final String CART_LINK       = ".shopping_cart_link";
    private static final String SORT_DROPDOWN   = ".product_sort_container";
    private static final String BURGER_MENU     = "#react-burger-menu-btn";
    private static final String LOGOUT_LINK     = "#logout_sidebar_link";

    public InventoryPage(Page page) {
        super(page);
    }

    public boolean isLoaded() {
        return locate(INVENTORY_LIST).isVisible();
    }

    public int getProductCount() {
        return locate(ITEM_NAMES).count();
    }

    public String getProductNameByIndex(int index) {
        return locate(ITEM_NAMES).nth(index).textContent();
    }

    public List<String> getAllProductNames() {
        return locate(ITEM_NAMES).allTextContents();
    }

    public void addToCartByIndex(int index) {
        locate(ADD_TO_CART_BTN).nth(index).click();
    }

    public void removeFromCartByIndex(int index) {
        locate(REMOVE_BTN).nth(index).click();
    }

    public int getCartBadgeCount() {
        if (!locate(CART_BADGE).isVisible()) return 0;
        return Integer.parseInt(locate(CART_BADGE).textContent().trim());
    }

    public CartPage goToCart() {
        locate(CART_LINK).click();
        return new CartPage(page);
    }

    public void selectSortOption(String value) {
        page.selectOption(SORT_DROPDOWN, value);
    }

    public LoginPage logout() {
        locate(BURGER_MENU).click();
        waitForVisible(LOGOUT_LINK);
        locate(LOGOUT_LINK).click();
        return new LoginPage(page);
    }
}

File: CSharp/Selenium.WebTests/Pages/InventoryPage.cs
using OpenQA.Selenium;

namespace Selenium.WebTests.Pages;

public class InventoryPage : BasePage
{
    private static readonly By InventoryList   = By.ClassName("inventory_list");
    private static readonly By InventoryItems  = By.ClassName("inventory_item");
    private static readonly By CartBadge       = By.ClassName("shopping_cart_badge");
    private static readonly By CartLink        = By.ClassName("shopping_cart_link");
    private static readonly By BurgerMenu      = By.Id("react-burger-menu-btn");
    private static readonly By LogoutLink      = By.Id("logout_sidebar_link");
    private static readonly By SortDropdown    = By.ClassName("product_sort_container");

    public InventoryPage(IWebDriver driver) : base(driver) { }

    public bool IsLoaded => IsElementVisible(InventoryList);

    public int ProductCount =>
        Driver.FindElements(InventoryItems).Count;

    public int CartItemCount
    {
        get
        {
            try { return int.Parse(Driver.FindElement(CartBadge).Text); }
            catch { return 0; }
        }
    }

    public void AddProductToCartByIndex(int index = 0)
    {
        var buttons = Wait.Until(d =>
        {
            var b = d.FindElements(By.CssSelector(".inventory_item button[id^='add-to-cart']"));
            return b.Count > index ? b : null;
        })!;

        // product slug: "add-to-cart-sauce-labs-backpack" → "sauce-labs-backpack"
        string slug = buttons[index].GetAttribute("id").Replace("add-to-cart-", "");
        ((OpenQA.Selenium.IJavaScriptExecutor)Driver).ExecuteScript("arguments[0].click();", buttons[index]);

        // Wait for the matching Remove button to appear (confirms the click fired).
        Wait.Until(d => d.FindElements(By.Id($"remove-{slug}")).Count > 0);
    }

    public void RemoveProductFromCartByIndex(int index = 0)
    {
        var buttons = Wait.Until(d =>
        {
            var b = d.FindElements(By.CssSelector(".inventory_item button[id^='remove']"));
            return b.Count > index ? b : null;
        })!;

        string slug = buttons[index].GetAttribute("id").Replace("remove-", "");
        ((OpenQA.Selenium.IJavaScriptExecutor)Driver).ExecuteScript("arguments[0].click();", buttons[index]);

        // Wait for the Add-to-cart button to re-appear, confirming removal fired.
        Wait.Until(d => d.FindElements(By.Id($"add-to-cart-{slug}")).Count > 0);
    }

    public string GetProductNameByIndex(int index = 0) =>
        Driver.FindElements(By.ClassName("inventory_item_name"))[index].Text;

    public CartPage GoToCart()
    {
        WaitForElement(CartLink).Click();
        Wait.Until(d => d.Url.Contains("cart"));
        return new CartPage(Driver);
    }

    public void SelectSortOption(string optionText)
    {
        var select = new OpenQA.Selenium.Support.UI.SelectElement(
            WaitForElement(SortDropdown));
        select.SelectByText(optionText);
    }

    public LoginPage Logout()
    {
        WaitForElement(BurgerMenu).Click();
        // SauceDemo's sidebar slides in via CSS animation; WaitForClickable isn't
        // enough — the element is "clickable" before the transition fully settles.
        // A JS click bypasses the animation overlay reliably.
        var logoutLink = WaitForClickable(LogoutLink);
        ((OpenQA.Selenium.IJavaScriptExecutor)Driver).ExecuteScript("arguments[0].click();", logoutLink);
        Wait.Until(d => d.Url == "https://www.saucedemo.com/");
        return new LoginPage(Driver);
    }
}
